config/consul.go

```diff
@@ -0,0 +1,655 @@
+
+	return true, nil
+}
+
+// List all apps in an environment
+func (c *ConsulBackend) ListApps(env string) ([]App, error) {
+	key := path.Join("galaxy", "apps", env)
+	kvPairs, _, err := c.client.KV().List(key, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	apps := make([]App, len(kvPairs))
+	for i, kvp := range kvPairs {
+		ad := &AppDefinition{}
+		err := json.Unmarshal(kvp.Value, ad)
+		if err != nil {
+			log.Println("error decoding AppDefinition for %s: %s", kvp.Key, err.Error())
+			continue
+		}
+		ad.ConfigIndex = int64(kvp.ModifyIndex)
+		apps[i] = App(ad)
+	}
+	return apps, nil
+}
+
+// Retrieve the current config for an application
+func (c *ConsulBackend) GetApp(app, env string) (App, error) {
+	key := path.Join("galaxy", "apps", env, app)
+	kvp, _, err := c.client.KV().Get(key, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if kvp == nil {
+		return nil, UnknownApp
+	}
+
+	ad := &AppDefinition{}
+	err = json.Unmarshal(kvp.Value, ad)
+	if err != nil {
+		return nil, err
+	}
+
+	ad.ConfigIndex = int64(kvp.ModifyIndex)
+	return ad, nil
+}
+
+// Update the current configuration for an app
+func (c *ConsulBackend) UpdateApp(app App, env string) (bool, error) {
+	ad := app.(*AppDefinition)
+	key := path.Join("galaxy", "apps", env, ad.Name())
+	kvp := &consul.KVPair{Key: key}
+
+	var err error
+	kvp.Value, err = json.Marshal(ad)
+	if err != nil {
+		return false, err
+	}
+
+	_, err = c.client.KV().Put(kvp, nil)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// Delete the configuration for an app
+// FIXME: Why does this take an App? Everything else takes a string
+func (c *ConsulBackend) DeleteApp(app App, env string) (bool, error) {
+	key := path.Join("galaxy", "apps", env, app.Name())
+	_, err := c.client.KV().Delete(key, nil)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// Add a pool assignment for this app, and update the config.
+// The pool need not exist, it just won't run until there is a corresponding
+// pool.
+func (c *ConsulBackend) AssignApp(app, env, pool string) (bool, error) {
+	appCfg, err := c.GetApp(app, env)
+	if err != nil {
+		return false, err
+	}
+
+	ad := appCfg.(*AppDefinition)
+	// return early if we're already assigned to this pool
+	for i := range ad.Assignments {
+		if ad.Assignments[i].Pool == pool {
+			return true, nil
+		}
+	}
+
+	// FIXME: we have to hard-code -1 here to match old behavior
+	a := AppAssignment{Pool: pool, Instances: -1}
+	ad.Assignments = append(ad.Assignments, a)
+	return c.UpdateApp(ad, env)
+}
+
+// Remove a pool assignment for this app, and update the config
+func (c *ConsulBackend) UnassignApp(app, env, pool string) (bool, error) {
+	found := false
+	appCfg, err := c.GetApp(app, env)
+	if err != nil {
+		return false, err
+	}
+
+	ad := appCfg.(*AppDefinition)
+	for i := range ad.Assignments {
+		if ad.Assignments[i].Pool == pool {
+			// remove the item for the slice
+			a := ad.Assignments
+			a[i], a = a[len(a)-1], a[:len(a)-1]
```

Comment:
This doesn't look quite right to me.  If `a` were `{"foo", "bar", "baz", "quux"}` I think this results in `a` being `{"foo", "quux", "bar", "quux"}` -- the last element isn't sliced off.